Strengthen local_safe_pointers to handle common Java operations

smowton opened the pull request:
The local-safe-pointers analysis can already inform symex that certain pointers
are known not to be null at particular program points. This strengthens the analysis
to spot more cases such that it can determine no null pointer is dereferenced in the
new regression test `jbmc/clean_derefs`, which exercises many common Java operations
including array accesses.

The specific improvements to local-safe-pointers:
 * Look through typecasts. This was already done for GOTO instructions, but now works
   for ASSUME as well.
 * Search for not-null expressions using base_type_eq instead of just irept::operator<.
   This means that when symex uses namespacet::follow to remove a symbol_typet that does
   not prevent local-safe-pointers from noting it is not null in a particular context.

smowton committed:
Strengthen local_safe_pointers to handle common Java operations

The local-safe-pointers analysis can already inform symex that certain pointers
are known not to be null at particular program points. This strengthens the analysis
to spot more cases such that it can determine no null pointer is dereferenced in the
new regression test `jbmc/clean_derefs`, which exercises many common Java operations
including array accesses.

The specific improvements to local-safe-pointers:
 * Look through typecasts. This was already done for GOTO instructions, but now works
   for ASSUME as well.
 * Search for not-null expressions using base_type_eq instead of just irept::operator<.
   This means that when symex uses namespacet::follow to remove a symbol_typet that does
   not prevent local-safe-pointers from noting it is not null in a particular context.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2f38d732b0f39f9d1ac42ea9dd3c0d717b5a9c59).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/79478407

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -13,66 +13,61 @@ Author: Diffblue Ltd
 
 #include <util/expr_iterator.h>
 #include <util/format_expr.h>
+#include <util/base_type.h>
```
Please maintain lexicographic sorting of includes.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -13,66 +13,61 @@ Author: Diffblue Ltd
+  exprt normalized_expr = expr;
   bool checked_when_taken = true;
-  while(normalized_guard.id() == ID_not || normalized_guard.id() == ID_equal)
+  while(normalized_expr.id() == ID_not || normalized_expr.id() == ID_equal)
```
I realise I should have called this out in the initial version of this code, but anyway: Could you please add a comment explaining what the purpose of this "normalisation step" is? As far as I can tell it would turn both `!x` and `!!x` into `x`, so all negations are removed. (But `checked_when_taken` may account for this.) Afterwards we only care about `ID_notequal`, meaning that really we only care about `a == b`, `a != b`, and any number of negations prefixing those two expressions. Could you please refactor this code to first do all the skipping of negations?

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -13,66 +13,61 @@ Author: Diffblue Ltd
+    while(op0->id() == ID_typecast)
+      op0 = &op0->op0();
+    while(op1->id() == ID_typecast)
+      op1 = &op1->op0();
```
I'd suggest using `skip_typecast` for both of the above.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -13,66 +13,61 @@ Author: Diffblue Ltd
+    if(op0->type().id() == ID_pointer &&
+       *op0 == null_pointer_exprt(to_pointer_type(op0->type())))
+    {
+      std::swap(op0, op1);
```
Maybe just do the same `return` as you do below (just for `op1` this time)?

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -82,7 +77,8 @@ get_conditional_checked_expr(const exprt &goto_guard)
 {
-  std::set<exprt> checked_expressions;
+  std::set<exprt, base_type_comparet> checked_expressions(
+    base_type_comparet{ns});
```
We don't usually use the `{ ... }` syntax.

smowton replied:
Yeah, but otherwise I need `(())` to disambiguate from a function declaration.

tautschnig commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -91,11 +87,23 @@ void local_safe_pointerst::operator()(const goto_programt &goto_program)
+      else
+      {
+        checked_expressions =
+          std::set<exprt, base_type_comparet>(base_type_comparet{ns});
```
As above

tautschnig reviewed (approved):
Looks ok to me except as noted below.

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2152,6 +2152,13 @@ static bool is_valid_string_constraint(
+    // constraint as s is indexed with both 'x' and 'x+1' under the same
+    // universal quantifier. We should either indirect to avoid the double
+    // reference, or establish that this conditon wasn't needed to begin with.
+    /*
```
Use `#if 0`

peterschrammel replied:
No sure I understand "produces a constraint, which does not meet this constraint".

allredj replied:
I had a think about that. A condition of the PASS algorithm is for the comparison predicate to be non-self-recursive.
Since the algorithm stops on a fix point (i.e. when no more indices are added to the index set), it might happen that adding an index will always result in also adding a shifted index (e.g. in the `s[x] == s[x+1]` case). So the algorithm never reaches a fix point and will enter an infinite loop (in the case of unbounded string lengths).

However, there's hope that early termination will happen anyway because of another mechanism. So in practice that might actually not happen often (or at all).

I would prefer the indirection solution, but if it's too complicated I can live with the pragmatic solution of removing the assertion and seeing what happens.

smowton committed:
Disable broken string test

This currently breaks an assertion as described in the .desc file. It should be re-enabled
once string-refinement knows how to introduce sufficient indirection to overcome this.

smowton commented:
Test-gen bump https://github.com/diffblue/test-gen/pull/2099

smowton commented:
@peterschrammel @allredj I've dropped the assertion disabling here and instead just marked the test KNOWNBUG; the strings team can take it as they wish from here :)

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: db303662679bc68cd723cc102d8f10990d05401b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80003479

peterschrammel commented on src/analyses/local_safe_pointers.cpp:
```diff
@@ -11,68 +11,68 @@ Author: Diffblue Ltd
+
+  if(normalized_expr.id() == ID_notequal)
+  {
+    const exprt *op0 = &skip_typecast(normalized_expr.op0());
```
Why a pointer?

smowton commented:
@peterschrammel from old code, fixed

peterschrammel commented on jbmc/regression/jbmc-strings/StartsWith/test_nondet.desc:
```diff
@@ -1,8 +1,13 @@
 assertion at file Test.java line 69 .*: FAILURE
 --
+--
+The nondet test currently breaks the invariant at string_refinement.cpp:2162
```
@allredj, can you please create a JIRA ticket for that?

peterschrammel commented on jbmc/regression/jbmc/clean_derefs/Test.java:
```diff
@@ -0,0 +1,50 @@
+
```
unnecessary blank line

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a0e339d57371255d5df1e4a2786250c03f687948).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80121763
